src/calendarworker.cpp

```diff
@@ -715,30 +716,29 @@ void CalendarWorker::setNotebookColor(const QString &notebookUid, const QString
 QHash<QString, CalendarData::EventOccurrence>
 CalendarWorker::eventOccurrences(const QList<CalendarData::Range> &ranges) const
 {
-    mKCal::ExtendedCalendar::ExpandedIncidenceList events;
-    foreach (CalendarData::Range range, ranges) {
-        // mkcal fails to consider all day event end time inclusivity on this, add -1 days to start date
-        mKCal::ExtendedCalendar::ExpandedIncidenceList newEvents =
-                mCalendar->rawExpandedEvents(range.first.addDays(-1), range.second,
-                                             false, false, QTimeZone::systemTimeZone());
-        events = events << newEvents;
-    }
-
-    QStringList excluded = excludedNotebooks();
+    const QStringList excluded = excludedNotebooks();
     QHash<QString, CalendarData::EventOccurrence> filtered;
-
-    for (int kk = 0; kk < events.count(); ++kk) {
-        // Filter out excluded notebooks
-        if (excluded.contains(mCalendar->notebook(events.at(kk).second)))
-            continue;
-
-        mKCal::ExtendedCalendar::ExpandedIncidence exp = events.at(kk);
-        CalendarData::EventOccurrence occurrence;
-        occurrence.eventUid = exp.second->uid();
-        occurrence.recurrenceId = exp.second->recurrenceId();
-        occurrence.startTime = exp.first.dtStart;
-        occurrence.endTime = exp.first.dtEnd;
-        filtered.insert(occurrence.getId(), occurrence);
+    foreach (CalendarData::Range range, ranges) {
+        KCalendarCore::OccurrenceIterator it(*mCalendar, QDateTime(range.first.addDays(-1)),
+                                             QDateTime(range.second.addDays(1)).addSecs(-1));
+        while (it.hasNext()) {
+            it.next();
+            if (mCalendar->isVisible(it.incidence())
+                && it.incidence()->type() == KCalendarCore::IncidenceBase::TypeEvent
+                && !excluded.contains(mCalendar->notebook(it.incidence()))) {
+                const QDateTime sdt = it.occurrenceStartDate();
+                const KCalendarCore::Duration elapsed
+                    (it.incidence()->dateTime(KCalendarCore::Incidence::RoleDisplayStart),
+                     it.incidence()->dateTime(KCalendarCore::Incidence::RoleDisplayEnd),
+                     KCalendarCore::Duration::Seconds);
+                CalendarData::EventOccurrence occurrence;
+                occurrence.eventUid = it.incidence()->uid();
+                occurrence.recurrenceId = it.incidence()->recurrenceId();
+                occurrence.startTime = sdt;
+                occurrence.endTime = elapsed.end(sdt);
```

Comment:
If there's a DST change in the middle, the endTime is not start+duration? Would that be a problem here?

Reply:
That's a good question. Here is a snippet from the old code from `mKCal`:
```
                    const QDateTime tsStartTime = kdatetimeAsTimeSpec(timeInInterval, tz);
                    const QDateTime tsEndTime = Duration(startTime, endTime).end(tsStartTime);
```
So at least, I'm not introducing any regression here.

That being said, could we do better ? Here are two situations that may arise:
1. we have a recurring event to bring our bat pet out from 1am to 3am, the bat requires 2 hours to feed.
2. we have a recurring event to work as a night guard from 22pm to 6am the next day.

In case of the first event, we're interested in the duration, so during a DST transition (assuming it is happening at 2am), the occurrence will span from 1am to 4am, which still correspond to a duration of 2 hours because of the DST making the time to be 3am when it was 2am. That's how it is coded now. On the contrary, for the second case, we're interested in starting and ending times. With the current coding strategy, the event will span up to 7am the day after, which would be wrong.

Internally, an event can be defined as a duration or with an end time. So we could make a if case here and apply a duration shift if the event is given as a duration or play with the date of the ending time to keep the hour part (could be tricky). But I think we have no way to decide if the event that was entered by the user was intended as a duration or from time to time, from a UI point of view (actually the UI is using the end time paradigm, but we often use it as a duration syntax).

So all in all, I would keep it like that since it does not break the current behaviour and is simple.

Reply:
> So at least, I'm not introducing any regression here.

That's a good and by such enough for me now :)  -> LGTM

> That being said, could we do better ?

You quite well listed the related problems. The incidences can be assumed either being based on time or duration. The UI could offer both, but then the user would need to be aware of the subtle differences. In theory nice to handle all cases but in practice likely rare case.